src/Servers/Kestrel/Core/src/Internal/Http2/AspNetCoreHPackEncoder.cs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http.HPack;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
+{
+    internal class AspNetCoreHPackEncoder
+    {
+        private HeaderEnumerator _enumerator;
+
+        private HPackEncoder _hPackEncoder = new HPackEncoder();
+
+        public bool BeginEncode(HeaderEnumerator enumerator, Span<byte> buffer, out int length)
+        {
+            _enumerator = enumerator;
+            _enumerator.MoveNext();
+
+            return Encode(buffer, out length);
+        }
+
+        public bool BeginEncode(int statusCode, HeaderEnumerator enumerator, Span<byte> buffer, out int length)
+        {
+            _enumerator = enumerator;
+            _enumerator.MoveNext();
+
+            int statusCodeLength = _hPackEncoder.EncodeStatusCode(statusCode, buffer);
+            bool done = Encode(buffer.Slice(statusCodeLength), throwIfNoneEncoded: false, out int headersLength);
+            length = statusCodeLength + headersLength;
+
+            return done;
+        }
+
+        public bool Encode(Span<byte> buffer, out int length)
+        {
+            return Encode(buffer, throwIfNoneEncoded: true, out length);
+        }
+
+        private bool Encode(Span<byte> buffer, bool throwIfNoneEncoded, out int length)
+        {
+            int currentLength = 0;
+            do
+            {
+                if (!_hPackEncoder.EncodeHeader(_enumerator.Current.Key, _enumerator.Current.Value, buffer.Slice(currentLength), out int headerLength))
+                {
+                    if (currentLength == 0 && throwIfNoneEncoded)
+                    {
+                        throw new HPackEncodingException("Failed to HPACK encode the headers.");
+                    }
+
+                    length = currentLength;
+                    return false;
+                }
+
+                currentLength += headerLength;
+            }
+            while (_enumerator.MoveNext());
+
+            length = currentLength;
+
+            return true;
+        }
+    }
+
+    internal struct HeaderEnumerator
```

Comment:
Make 2 different enumerators? One for trailers and one for headers? This struct ends up being bigger as a result of merging them.

Reply:
If there are two enumerators then there needs to be two copies of BeginEncode/Encode methods on AspNetCoreHPackEncoder, one for each struct.

I think a slightly larger struct it worth avoiding duplicating the struct and all the encoding logic.

Reply:
How big is the struct after this?

Reply:
232 bytes.

I've got an idea. Writing headers and trailers is always done in a lock. The enumerator could be a class that lives on Http2FrameWriter. When you want to use it you just reset it with either the headers or trialers collection you want to iterate over.

No big struct, no allocations. What do you think?